Numark Mixtrack Platinum midi mapping

iamcodemaker opened the pull request:
This is a mapping for the Numark Mixtrack Platinum controller. Nearly all of the features of the controller are supported including the LCD display screens and jog wheels. Support for controlling Mixxx's effects and effect parameters is supported as well. I have also extended the functionality of the controller to support 8 hotcues.

Documentation here https://www.mixxx.org/wiki/doku.php/numark_mixtrack_platinum

There are a bunch of commits here, I can squash if necessary.

iamcodemaker committed:
Numark Mixtrack Platinum: add mapping

This is a mapping for the Numark Mixtrack Platinum. Most features of
the controller are supported including the LCD screens and 4 deck
control. This mapping is targeted for Mixxx 2.0.0.

Notes:
- FX buttons always operat channels 1 and 2. This is a limitation of
the controller that could be scripted around, but I got lazy (also how
it works in Serato).
- Both FX dry/wet strips control the same effects chain and will affect
both channels 1 and 2 if effects are enabled on them. Serato has
isolated effects banks for each channel.
- The "BEATS" knobs are not mapped, this concept did not translate
cleanly to Mixxx.
- Sometimes there are artifacts in the auto after long spinback
maneuvers. This may not have anything to do with the controller.
- The sub-beat auto-loop buttons do not light up when enabled unless
the shift key is pressed. This is different from how the work in Serato.
- Master sync enable and disable works like in the Mixxx software interface,
not how it works in Serato. Shift-sync is not mapped.
- The slip mode indicator is not mapped, this is trivial to fix (midi
codes are documented in the JS file), but I couldn't figure out how to
use slip mode in the Mixxx interface.
- The up/down indicators on the LCD screens to indicate which direction
to adjust the pitch slider are not implemented. It's possible to do, but
would require some scripting.
- Browse knob press is unmapped. I couldn't get this to do something
sensible in Mixxx 2.0.0. Version 2.1.0 seems to have more sensible
options.

iamcodemaker committed:
sync, play, cue

iamcodemaker committed:
hotcues

iamcodemaker committed:
pfl

iamcodemaker committed:
Numark Mixtrack Platinum: update schema compatibility

iamcodemaker committed:
exit demo mode first

iamcodemaker committed:
pitch

iamcodemaker committed:
decks go in array/component container

iamcodemaker committed:
sampler

iamcodemaker committed:
loop

iamcodemaker committed:
fixup: shutdown sequence

iamcodemaker committed:
loop double/half

iamcodemaker committed:
fixup: hotcues

iamcodemaker committed:
basic effects

iamcodemaker committed:
browse knob

iamcodemaker committed:
remove helper led init functions

iamcodemaker committed:
fixup effects

iamcodemaker committed:
effects dryWetKnob

iamcodemaker committed:
refactor effects section mappings

iamcodemaker committed:
tweak scratch settings to help the mark and spinner line up with the jog wheel position

iamcodemaker committed:
tweak scratch algorithm to add some forward momentum to the platter

iamcodemaker committed:
tweak scratching behavior

Behave differently when the platter is spinning forward vs backwards
but not being touched. Effectively this changes causes the track to
stop scratch mode and start playing sooner when moving foward vs
backwards. This allows better spinbacks, and more natural cueuing when
seeking backwards with the platter and for cleaner exits when
scratching when pushing the platter forward at the end of a scratch.

iamcodemaker committed:
tweak the way searching the track works

iamcodemaker committed:
remove "(local)" from the name

iamcodemaker committed:
use PeakIndicator to control the red section of the vu meter

Thanks JosepMa for this suggestion.

iamcodemaker committed:
remove duplicate mapping for channel1 rate_temp_up

Thanks JosepMa!

iamcodemaker committed:
remove stray `<key>` entry

Thanks JosepMa!

iamcodemaker committed:
add support for beat jumping (shift+jog)

Thanks JosepMa!

iamcodemaker committed:
use GoToItem instead of ChooseItem for browse knob press

iamcodemaker committed:
tweak scratch parameters for 2.1 rc1

iamcodemaker committed:
detect really fast platter spins properly

If the platter is spun very fast and let go, previously we would detect
this as the platter going in the opposite direction, now we detect this
properly.

iamcodemaker committed:
make the tap button work on decks 3 and 4

iamcodemaker committed:
use scripted components for EQ knobs

This will allows us to map these to effect parameters when shift is
pressed.

iamcodemaker committed:
make touch strip control effect super knob

Also only use one of the touch strip midi inputs instead of mapping all
three separately.

iamcodemaker committed:
implement focused effect controls

iamcodemaker committed:
re-order parameter knobs for focused effects

Order is now filter, gain, high, mid, low.

iamcodemaker committed:
fix touch strip behavior when focusing effects

iamcodemaker committed:
wait until we configure our controls before initializing the controller

iamcodemaker committed:
don't ignore initial touch of touch strip

iamcodemaker committed:
store information about active decks in the deck object

iamcodemaker committed:
use a function to activate and de-activate decks

This will be used to set up soft takeover for the pitch sliders.

iamcodemaker committed:
enable soft takeover on the pitch controls

iamcodemaker committed:
fix connect/disconnect for focused effects

This didn't appear to break anything but the ordering of things was
wrong.

iamcodemaker committed:
properly initialize deck and effect unit LEDs again

This broke on 569fd2991c5ab5b3255416c184e358f436274450, the one in
which we set up our controls before initializing the controller.

iamcodemaker committed:
add jshint header

iamcodemaker committed:
fix bug with searching using the jogs when wheel is disabled

iamcodemaker committed:
add config option for wheel mode

iamcodemaker committed:
add config for default state of elapsed/remaining display

iamcodemaker committed:
rename XML file to match naming convention

iamcodemaker committed:
use shift+cue gain to adjust the sampler volume

iamcodemaker committed:
use loop_scale to halve and double loops

iamcodemaker committed:
switch loop in/out controls to script bindings

Decks 2-4 were still using the normal xml bindings.

iamcodemaker committed:
add an option to use the manual loop controls as cue 5-8

iamcodemaker committed:
change decription to LCD screens from LED screens

iamcodemaker committed:
enable loop roll for sub 1 beat auto loops

This is how it works in serato.

iamcodemaker committed:
script auto loop controls and add hotcue option

This adds an option to map the auto loop buttons as hotcues 5-8.

iamcodemaker committed:
remove redundant auto loop led init code

iamcodemaker committed:
make shift+padmode switch the targeted padmode to control hotcues

iamcodemaker committed:
control sampler banks 5-8 when shift+padmode+sampler is pressed

Thanks Compino Foxx for the suggestion.

radusuciu commented:
Looks like a very comprehensive mapping. No glaring issues popped up during my reading of this code. I would recommend looking at this though: https://www.codefactor.io/repository/github/mixxxdj/mixxx/pull/1855. You can safely ignore the "variable declared not used" warnings imo, but you may want to think about the code duplication issues.

Minor nitpick, some mixing of camelCase with under_score, but I don't think this should hold up merge.

Has this been tested by others?

iamcodemaker committed:
use the hotcue buttons for sampler banks 5-8

When shift+padmode+sampler is pressed, the hotcues will operate sampler
slots 5-8.

iamcodemaker committed:
add UseCueAsSampler option

This swaps the shift+padmode and normal padmode behavior for the
sampler.

iamcodemaker committed:
add link to wiki in xml

Be-ing commented:
Thanks for submitting this @iamcodemaker! And thanks for the review comments @radusuciu. I quickly glanced at the code and it looks neat and tidy. I'll go through the wiki and review the design for this soon, but right now my main focus is on fixing bugs for the 2.2.0 release. If we don't get it all done by the 2.2.0 release it's no big deal because we can easily make point releases.

iamcodemaker committed:
remove unnecessary and duplicate code from EffectUnit

iamcodemaker committed:
always stop scratching when not touching the platter

Previously, if the wheel button was disabled while scratching, the
platter touch off event was ignored. Now it will be properly handled.

iamcodemaker committed:
don't use connect()/disconnect() to manage soft takeover

Manage soft takeover directly instead of manually calling connect() and
disconnect().

iamcodemaker committed:
only ignore next value if headgain/sampler control moves

Only ignore the next value when the headgain/sampler volume control
moves when shift is pressed.

iamcodemaker committed:
shift+filter controls meta knob

This is only enabled when no effect is focused.

iamcodemaker committed:
toggle keylock when both pitch buttons are pressed

iamcodemaker committed:
use shift+pitch bend to adjust key

Press both pitch bend keys to reset.

iamcodemaker commented:
I think I fixed the soft takeover issue on the EQ knobs. I'll need to test some more to be sure. Also there is another known issue: the first platter touch doesn't activate scratch mode properly for some reason. Things work fine on subsequent touches.

iamcodemaker committed:
make shift+load load and play (or eject)

Configurable with the ShiftLoadEjects config option.

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+            components.Button.prototype.input.call(this, channel, control, value, status, group);
+            if (this.isPress(channel, control, value, status)) {
+                eu.forEachComponent(function (component) {
+                    if (component.tap !== undefined) {
```
Not strictly necessary, but a good safety measure
```suggestion
                    if (component.tap !== undefined && typeof component.tap === 'function') {
```

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+        this.group = '[EffectRack1_EffectUnit' + eu.currentUnitNumber +
+                      '_Effect' + this.number + ']';
+        this.midi = [0x97 + eu.currentUnitNumber, this.number - 1];
+        this.inToggle = components.Button.prototype.inToggle;
```
Is this necessary?

iamcodemaker replied:
probably not, I'll investigate

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+        toggle_focused_effect: function() {
+            if (engine.getValue(eu.group, "focused_effect") === this.number) {
+                engine.setValue(eu.group, "show_focus", 0);
+                engine.setValue(eu.group, "show_parameters", 0);
```
I'm not sure this should toggle `show_parameters`. I would presume that mostly users with this controller would be using effect focusing to control the metaknob with the touchstrip, in which case showing all the parameters would be an annoying waste of space. On the other hand, I could imagine that a few users would be using the mixer knobs to adjust the individual parameters when an effect is focused. I'm a bit hesitant to add yet another option for this...

iamcodemaker replied:
I was on the fence about this. When I focus an effect, I use it to change the parameters. Though if you don't do that, I can see this being an issue. The only way around it would be adding an option. Side note: Maybe we need a UI for controller mapping options. That's something I'm willing to work on.

Be-ing replied:
Okay, let's make it an option then. I agree, we should have a GUI for controller mapping options. But before making major changes to the mapping system I think we should finish #1795.

iamcodemaker replied:
Updated. I left setting `show_focus` enabled unconditionally, what are your thoughts on that? Should that also fall under the option to show the parameters?

Be-ing replied:
Yes, `show_focus` should be toggled regardless of the option. There should not be a case where there is focus but the user can't see it.

iamcodemaker replied:
I believe there is one case, if you exit mixxx with an effect focused, that effect will be focused when you start again (the controller button will flash), but the ui will not show it as focused. I'd have to double check. This is something of a weird edge case though.

Be-ing replied:
Hmm, if you can reproduce it try unplugging your controller before starting Mixxx again. There may be something in the mapping causing that.

iamcodemaker replied:
Yup. Tested this again. If you exit with an effect focused, it remains focused when you start again, but the ui doesn't show it. Should I make the mapping check for this on startup and enable `show_focus`?

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+        shiftControl: true,
+        shiftOffset: 4,
+        shift: function() {
+            this.inKey = 'start';
```
Why not just use the default shift behavior of `components.CueButton`?

Be-ing replied:
When I look at the icon printed on the controller, I would expect the default `start_stop` behavior. I don't really understand the use case for not stopping playback with this. If you want to jump to the start and keep playing, you can use a normal hotcue.

> The reasoning here is that most people who get this controller and use it with mixxx won't be familar with mixxx.

I don't think that's a reliable assumption. Many Mixxx users do not have (much) experience with other DJ software, especially Linux users.

iamcodemaker replied:
I'm ok with going with the `start_stop` default. Thinking this through, jumping to the start and continuing to play doesn't seem very useful. Shift+play will jump to the cue point and I can't think of a reason you would want to jump to the start of the track instead of the cue point and keep playing. I always use this button to jump to the start when not playing the track, so making this the default would actually save me a button press.

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+    }
+
+    var pitch_or_keylock = function (channel, control, value, status, group) {
+        if (this.other.inGetValue() > 0.0 && this.isPress(channel, control, value, status)) {
```
Clever way of implementing this :+1:

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+    var pitch_or_keylock = function (channel, control, value, status, group) {
+        if (this.other.inGetValue() > 0.0 && this.isPress(channel, control, value, status)) {
+            // toggle keylock, both keys pressed
+            engine.setValue(this.group, "keylock", !engine.getValue(this.group, "keylock"));
```
You can use the `script.toggleControl` shorthand.
```suggestion
            script.toggleControl(this.group, "keylock");
```

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+                var old = this.inGetValue();
+                // the gui only allows 7 steps, we replicate that behavior here
+                if (Math.abs(old + this.direction) < 8.0) {
+                    this.inSetValue(Math.round(old + 1.0 * this.direction));
```
I think this is more complicated than it needs to be. Why not just use `pitch_up/down`?

Be-ing replied:
Oh, good catch. I think that is because the `_up` and `_down` ControlPushButtons are [automatically generated](https://github.com/mixxxdj/mixxx/blob/master/src/control/controlpotmeter.cpp#L60) from the [pitch ControlPotmeter](https://github.com/mixxxdj/mixxx/blob/master/src/engine/keycontrol.cpp#L24). In this case I think it would be helpful to specifically document them though. I only found them by looking at tooltips for the buttons in Deere in Developer Mode. Could you add them to the wiki?

iamcodemaker replied:
lol, was hoping you would do it. I'll take care of it.

Be-ing replied:
Thanks

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+            number: 8,
+        })),
+        loop_double: new components.HotcueButton(loop_base(0x35, {
+            number: 8,
```
There are two "hotcue button 8"s here. Did you accidentally do an extra copy & paste?

Be-ing replied:
I'm confused. That's what `sendShifted` is for and it's only hotcue 8 that has duplicate buttons here.

Be-ing replied:
Okay, please add a comment explaining this.

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+        loop_toggle: new components.LoopToggleButton(loop_base(0x32, {})),
+        loop_halve: new components.Button(loop_base(0x34, {
+            key: 'loop_halve',
+            input: function(channel, control, value, status) {
```
Why is this needed? Doesn't `loop_halve` with the default toggle Button behavior work? Likewise for `loop_double` below.

iamcodemaker replied:
It works, but it doesn't change the loop length for non-quantized loops. Setting `loop_scale` works for quantized and non-quantized loops. In fact the revision where `loop_halve` was changed to only work for quantized loops mentions this as a work around.

Be-ing replied:
Oh right, I forgot the context that this is in manual loop mode. Also, this mapping does not use `beatloop_toggle`, so updating `beatloop_size` doesn't really matter much. So I guess it is okay to leave it as is here.

Maybe we should introduce new ControlPushButtons for this use case instead of the unusual `loop_scale` Control. Figuring out how loop halve/double buttons should work when designing the current looping interface was very difficult and there may still be room for improvement. If you want to dig into the weeds of the discussion, it was in #1187.

Be-ing replied:
I added a note to the [wiki page](https://www.mixxx.org/wiki/doku.php/numark_mixtrack_platinum#manual_loop_mode) documenting this.

Be-ing reviewed (changes requested):
I haven't read through all the code yet, but here's a start.

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+        // update duration, time elapsed, and the spinner
+        var duration = engine.getValue(group, 'duration');
+        var position = engine.getValue(group, 'playposition');
+        if (duration == 0) position = 0;
```
An artifact of my personal coding style. For single line if statements I put everything on one line and omit the brackets. Being on one line makes it harder to accidentally add another statement with the same indent but leaving it out the if statement. I'll update it (and other instances) according to your style.

Be-ing replied:
Always doing it one way means you don't have to worry about such subtle bugs and quirks of the language :)

iamcodemaker replied:
I'll make a sweep through and expand these soon.

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+
+MixtrackPlatinum.shutdown = function() {
+    // note: not all of this appears to be strictly necessary, things work fine
+    // with out this, but other software has been observed sending these led
```
Serato. Will update.

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+    // turn off a bunch of channel specific leds
+    for (var i = 0; i < 4; ++i) {
+        // pfl/cue button leds
+        midi.sendShortMsg(0x90 | i, 0x1B, 0x01);
```
Initially when I wrote this, I don't think the component system existed (or I wasn't aware of it). This code was copied over from that implementation. Some of these don't have components associated with them. It would be somewhat of a hassle to go through and figure out what's what and I don't see any benefit to it.

Be-ing replied:
Okay, leave as it is then

Be-ing committed:
Apply suggestions from code review

Co-Authored-By: iamcodemaker <matt@matt-land.com>

iamcodemaker committed:
fix typo in UseAutoLoopAsCue

iamcodemaker committed:
simplify key change buttons

iamcodemaker committed:
make reference to Serato more specific

iamcodemaker committed:
make shift+cue jump to start and stop the track

iamcodemaker committed:
remove unnecessary code

iamcodemaker committed:
make showing effect parameters optional

iamcodemaker committed:
use makeConnection instead of connectControl

iamcodemaker committed:
explain the duplicate hotcue 8 manual loop controls

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+    // note: not all of this appears to be strictly necessary, things work fine
+    // with out this, but other software has been observed sending these led
+    // reset messages during shutdown. The last sysex message may be necessary
+    // to re-enable demo mode.
```
> Users could also just unplug the controller when they are done

Well, that's what I found so annoying with the Hercules P32, which has no way to disable the light show. I had to go through the extra step of unplugging it when I was just going to leave it sitting on my desk. That also made it quite annoying to quickly test something in Mixxx.

iamcodemaker replied:
Going to leave this as is. I don't think it's a major concern for most users.

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1418 @@
+        })),
+    });
+    // swap normal and alternate manual loop controls
+    if (UseManualLoopAsCue) {
```
I figure most users won't even know about the extra options and in that case, the controller will work very similar to how the manual it comes with says it should. I tried to make all the defaults unsurprising. But when I play, I rarely use the autoloop functionality, so for me it makes sense to make that control cue points by default.

For the sampler, it probably makes sense to control 8 banks by default, but this might be unexpected to a user, so I want them to opt into it explicitly.

iamcodemaker replied:
Going to leave as is. This will be less of an issue when we have a UI for controller options. The code will be just as complex, but the burden on the user will be less. I still think having all the options available is useful for users who want to customize their workflow.

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1418 @@
+    MixtrackPlatinum.decks[3] = new MixtrackPlatinum.Deck(3, 0x02, MixtrackPlatinum.effects[1]);
+    MixtrackPlatinum.decks[4] = new MixtrackPlatinum.Deck(4, 0x03, MixtrackPlatinum.effects[2]);
+
+    // set up two banks of samplers, 4 samplers each
```
Unfortunately, there is no way to do this configuration. The controller sends the same midi codes regardless of the deck you are using to activate the sampler pads. We can tell when you switch into sampler mode on a deck, but the button presses aren't associated with the deck.

Be-ing replied:
:man_facepalming: :man_facepalming: why do manufacturers do these things...

Be-ing replied:
Noted on the wiki https://www.mixxx.org/wiki/doku.php/numark_mixtrack_platinum#sample_mode

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1418 @@
+MixtrackPlatinum.scratch_accumulator[2] = 0;
+MixtrackPlatinum.scratch_accumulator[3] = 0;
+MixtrackPlatinum.scratch_accumulator[4] = 0;
+MixtrackPlatinum.last_scratch_tick = [];
```
Should this array be initialized with 0s?

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1418 @@
+    inKey: 'headGain',
+    input: function (channel, control, value, status, group) {
+        if (this.ignore_next == "sampler" && !this.shifted) {
+            this.sampler.forEachComponent(function(component) {
```
I explained this above, but in short, doing this in `shift/unshift` can cause a "phantom" soft takeover mode. `ignoreNextValue()` gets called by `disconnect()` even if the control in question never moved every time shift is pressed. This strategy causes `ignoreNextValue()` to only get called if the control moved.

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1418 @@
+            var focused_effect = engine.getValue(eu.group, "focused_effect");
+            if (focused_effect === 0) {
+                if (this.shift_key) {
+                    engine.softTakeover(eu.group, this.shift_key, true);
```
Okay, that's what I thought the issue might be. That might explain some issues I was having trouble reproducing when using my Hercules P32 mapping. So the implementation of the generic `Pot` in the library should be fixed.

Be-ing replied:
I updated the [scripting wiki page](https://mixxx.org/wiki/doku.php/midi_scripting#soft-takeover) to document this. Thanks for pointing it out.

Be-ing replied:
I noticed a related [bug in soft takeover](https://bugs.launchpad.net/mixxx/+bug/1805049).

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1418 @@
+MixtrackPlatinum.screenBpm = function(deck, bpm) {
+    bpm = MixtrackPlatinum.encodeNum(bpm);
+    bpm.shift();
+    bpm.shift();
```
Oh, I got confused because I misread the code and thought the `MixtrackPlatinum.encodeNum` function was being assigned instead of invoked.

iamcodemaker committed:
init scratch tracking arrays

iamcodemaker committed:
update comment

iamcodemaker committed:
rename scratchTimer to scratchTimerCallback

Also use a closure instead of the legacy text method to set the callback
function.

iamcodemaker committed:
remove redundant inKey setting for BrowseKnob

iamcodemaker committed:
add notes about softTakeoverIgnoreNextValue()

iamcodemaker committed:
rename encodeNum to encodeNumToArray

iamcodemaker committed:
update comment

iamcodemaker committed:
update comment

iamcodemaker committed:
don't reuse bpm variable name

iamcodemaker committed:
make the code in positionCallback() clearer

iamcodemaker committed:
tweak comment

iamcodemaker committed:
link elapsed/remaining time to the value shown in the gui

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1467 @@
+
+    // update the time display
+    var time = MixtrackPlatinum.timeMs(midi_chan + 1, playposition, duration);
+    MixtrackPlatinum.screenTime(midi_chan + 1, time);
```
How does this handle showing time remaining? Or do you just need to send the controller firmware the time elapsed and duration, then it calculates remaining if you set it to show remaining?

iamcodemaker replied:
Right. We tell the controller the duration and we send the current time elapsed. The controller shows remaining/elapsed depending on which one it has been told to do.

Be-ing replied:
Okay, please add a comment explaining that.

Be-ing commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1418 @@
+    midi.sendShortMsg(0xBF, 0x45, 0);
+
+    // setup position tracking
+    engine.makeConnection("[Channel1]", "playposition", MixtrackPlatinum.positionCallback);
```
ping

iamcodemaker replied:
Haven't forgotten about this.

iamcodemaker committed:
relocate deckswitch and vu metering code

iamcodemaker committed:
remove random unused MixtrackPlatinum.pitch array

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1467 @@
+    //
+    // the visual spinner in the mixxx interface takes 1.8 seconds to loop
+    // (60 seconds/min divided by 33 1/3 revolutions per min)
+    var period = 60 / (33+1/3);
```
I'll make this change, but I actually think all this indirection makes the math harder to read.

iamcodemaker committed:
break out position indicator formula

iamcodemaker committed:
properly show focus when an effect is focused

Previously, this would only take effect when triggered by the controller
buttons. Now any focus change (like at mixxx startup) will triger
show_focus.

iamcodemaker committed:
handle bpm in the deck object

iamcodemaker committed:
move position tracking into the deck object

iamcodemaker committed:
use a component for bpm control

iamcodemaker committed:
use a component to track position

This is necessary so that when we trigger the entire deck, the position
gets updated properly. This happens after sending the controller init
message which will clear out the position value, and thus needs to be
re-sent.

iamcodemaker committed:
expand if statement

iamcodemaker committed:
use a control for tracking duration

iamcodemaker committed:
rename screen update functions

iamcodemaker committed:
rename array vars in screen update functions

iamcodemaker committed:
make sure all arrays are initialized before use

iamcodemaker commented on res/controllers/Numark-Mixtrack-Platinum-scripts.js:
```diff
@@ -0,0 +1,1415 @@
+        if (engine.getValue(group, key)) {
+            midi.sendShortMsg(0x90 | midi_channel, midino, 0x7F);
+        }
+        else {
```
I'll make a sweep and fix these soon.

iamcodemaker committed:
initialize the scratch_direction array

iamcodemaker committed:
when not touching the platter, stop scratching sooner

When spinning the platter backwards and letting it go, stop scratching
sooner. This enables tighter scratching.

Be-ing commented:
I was about to merge this but I noticed it was targeted at master. Can you rebase this on 2.2 before merging?

I had a chance to play with a Mixtrack Platinum with Serato for a little bit a few weeks ago and I think it's a great deal and a good controller to recommend for new users. Thanks for your work on this!

iamcodemaker commented:
Rebased. I was still planing on going through and tweaking the formatting of if/else statements and such, though it may be a few days before I can get to that. Obviously you can merge with out that.
